### PR TITLE
Feat/talk2scholars mainrouter update - Migration to LangGraph Hierarchical Agent System

### PR DESCRIPTION
**Pull request by ansh-info**
# For authors

https://github.com/user-attachments/assets/18cf1746-5e5f-4205-8968-4f284008e481

## Description

This PR fully transitions the agent system to a hierarchical structure based on [LangGraph’s official multi-agent tutorial](https://langchain-ai.github.io/langgraph/tutorials/multi_agent/hierarchical_agent_teams/) 

Introduced a router in the main agent to dynamically route queries to sub-agents.  

- Implements a routing system to invoke a sub-agent (e.g., `s2_agent`) or generate a direct response.  
- Uses a `Router` to determine the next step.  
- Added `call_s2_agent` to invoke the Semantic Scholar agent and update the conversation state.  
- For `StateGraph` now we can just add another node when we introduce new agents(benefit of this approach).

The addition display and querying tools in the agent system, improving how papers are displayed and queried after a search or recommendation.  

#### Key Updates  

- `display_results` Tool:  
  - Retrieves and displays the current/last set of papers found.  
  - Returns results as an artifact to be displayed in the frontend.  
  - Raises `NoPapersFoundError` if no papers are available, and trigger the search(if the search result was empty).  

- `query_results` Tool:  
  - Enables querying of the last displayed papers using this tool.  
  - Converts retrieved papers into a Pandas DataFrame for queries.  
  - Uses `create_pandas_dataframe_agent` from LangChain Experimental, which:  
    - Creates an agent capable of interacting with tabular data in Pandas.  
    - Supports natural language queries over a DataFrame.  
  - Raises `NoPapersFoundError` if no papers have been displayed yet.  

#### State Management & Future Work  

- We have a key in the state called `last_displayed_paper`, which keeps appending the results each time papers are retrieved.  
- This setup will allow querying not just the last displayed paper but also previously fetched papers, ensuring a broader context for recommendations.  
- This enables user to be able to query not just the last displayed papers but also the ones before 

---

#### Tools
| Tool | Purpose |
|----------|------------|
| `display_results.py` | Displays search and recommendation results in the form of dataframe in the front end. |
| `query_results.py` | User can call this tool to query to the selcted paper(using abstarct for now, later we will be integrating agent pdf so this will be helpful then). |
| `retrieve_semantic_scholar_paper.py` | Retrieves Semantic Scholar IDs based on paper titles. This tool will be used for Agent Zotero (which is currently under development and will be added in a subsequent PR). Since the papers fetched from the Zotero library will only have their titles, this tool allows us to retrieve the Semantic Scholar ID of a selected paper and then search for it using the search.py tool. |

---

### 1. Main agent routing
-  Added a router class for routing supervisor that dynamically selects sub-agents.  
- All sub-agents operate independently, following LangGraph’s hierarchical design.  

#### 3. Keys State Management
- Added `last_displayed_papers` key → Enables the LLM to answer queries based on previously retrieved papers, since we are returning artifacts and updating the state, the llm won't have access to the papers now, and now we capture the artifacts in the frontend and display raw results
- Added `llm_model` key into `shared_state` that uses Basechatmodel which is a centralized approach and this can be passed on to other sub agents

##### Additional Enhancements

- Tables are now directly displayed in the frontend using artifacts in form of dataframe and are diplayed in the front end using `display_results.py` 
- Streamlit now integrates Hydra config 

## Fixes # (issue) Mention the issue number.
#76 #97 

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [x] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Review by dmccloskey — CHANGES_REQUESTED** [review]
The migration is looking good 🤩. Mostly just minor comments.

File: aiagents4pharma/talk2scholars/agents/main_agent.py [hunk 1]
```diff
@@ -54,116 +43,134 @@ def get_hydra_config():
...
+        messages = [SystemMessage(content=router_prompt)] + state["messages"]
+        structured_llm = llm_model.with_structured_output(Router)
+        response = structured_llm.invoke(messages)
+        goto = response.next
+        # if "next" in response:
+        #     goto = response["next"]
+        # else:
+        #     goto = response["properties"]["next"]
+        logger.info("Routing to: %s, Thread ID: %s", goto, thread_id)
+        print("GOTO: ", goto)
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/main_agent.py:103]: Same for the print statements, it was added during the testing

File: aiagents4pharma/talk2scholars/agents/s2_agent.py [hunk 2]
```diff
@@ -46,31 +89,35 @@ def agent_s2_node(state: Talk2Scholars):
         cfg = cfg.agents.talk2scholars.s2_agent
 
     # Define the tools
-    tools = ToolNode([s2_search, s2_display, s2_single_rec, s2_multi_rec])
+    # tools = ToolNode([s2_search, s2_display, s2_single_rec, s2_multi_rec])
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/s2_agent.py:92]: No its not needed, I deleted it, kept it during testing(these were the old tools)

File: aiagents4pharma/talk2scholars/agents/s2_agent.py [hunk 3]
```diff
@@ -46,31 +89,35 @@ def agent_s2_node(state: Talk2Scholars):
...
     logger.log(logging.INFO, "Using OpenAI model %s", llm_model)
-    llm = ChatOpenAI(model=llm_model, temperature=cfg.temperature)
 
     # Create the agent
     model = create_react_agent(
-        llm,
+        llm_model,
         tools=tools,
         state_schema=Talk2Scholars,
         # prompt=cfg.s2_agent,
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/s2_agent.py:112]: I removed it, thank you

File: aiagents4pharma/talk2scholars/agents/s2_agent.py [hunk 4]
```diff
@@ -46,31 +89,35 @@ def agent_s2_node(state: Talk2Scholars):
...
 
-    # Define a new graph
     workflow = StateGraph(Talk2Scholars)
-
-    # Define the two nodes we will cycle between
     workflow.add_node("agent_s2", agent_s2_node)
-
-    # Set the entrypoint as `agent`
-    # This means that this node is the first one called
     workflow.add_edge(START, "agent_s2")
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/s2_agent.py:119]: The current edge is correct, since now we moved to Router, we don't need the commented edge, I removed it.

File: aiagents4pharma/talk2scholars/configs/agents/talk2scholars/main_agent/default.yaml
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/agents/talk2scholars/main_agent/default.yaml:1]: Sorry for this mistake, I’ve reviewed the spelling and fixed a few typos, such as ‘semactic’ → ‘semantic’ and ‘dont’ → ‘don’t.’ I also standardized the terminology to ensure consistency (e.g., ‘Semantic Scholar ID’). Let me know if you spot anything else!

File: aiagents4pharma/talk2scholars/configs/tools/multi_paper_recommendation/default.yaml [hunk 5]
```diff
@@ -9,6 +9,7 @@ api_fields:
   - "authors"
   - "citationCount"
   - "url"
+  # - "externalIds"
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/tools/multi_paper_recommendation/default.yaml:12]: Yes, this is the param to fetch arXiv id for a paper, Its working I have commented it out right now we will use it when we integrate agent pdf, I have added the comment "# Commented fields that could be added later if needed"

File: aiagents4pharma/talk2scholars/configs/tools/search/default.yaml [hunk 6]
```diff
@@ -9,6 +9,7 @@ api_fields:
   - "authors"
   - "citationCount"
   - "url"
+  # - "externalIds"
 # Commented fields that could be added later if needed
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/tools/search/default.yaml:13]: Thank you

File: aiagents4pharma/talk2scholars/tools/s2/multi_paper_rec.py [hunk 7]
```diff
@@ -101,26 +104,50 @@ def get_multi_paper_recommendations(
...
+                        tool_call_id=tool_call_id,
+                    )
+                ]
+            }
+        )
+
     # Create a dictionary to store the papers
     filtered_papers = {
         paper["paperId"]: {
+            # "semantic_scholar_id": paper["paperId"],  # Store Semantic Scholar ID
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/s2/multi_paper_rec.py:122]: Yes, this will be implemented in the next pr where I will add zotero, and will change the argument for rec tools from `paper_ids` to `semantic_scholar_id`, since we have added arXiv id, the llm was getting confused which id to pick, so for the mean time I hashed it.

File: aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py [hunk 8]
```diff
@@ -90,32 +93,59 @@ def get_single_paper_recommendations(
...
+                    ToolMessage(
+                        content=f"No recommendations found for {paper_id}.",
+                        tool_call_id=tool_call_id,
+                    )
+                ],
+            }
+        )
+
     # Extract paper ID and title from recommendations
     filtered_papers = {
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py:118]: Thanks for the feedback! I understand the importance of creating a utility function to improve maintainability and reduce redundancy. I’ve already implemented the Zotero read tool following a similar pattern, and I plan to introduce the Zotero tool in my next PR. Given this, I will try to create this utility function in the next PR—if not, then in a future PR. To track this, I’ve already created an issue for it. Let me know if that sounds good!

File: app/frontend/streamlit_app_talk2scholars.py [hunk 9]
```diff
@@ -133,20 +126,16 @@ def update_llm_model():
...
-        # LLM panel (Only at the front-end for now)
-        llms = ["gpt-4o-mini", "gpt-4-turbo", "gpt-3.5-turbo"]
-        # llm_option = st.selectbox(
-        #     "Pick an LLM to power the agent",
-        #     llms,
-        #     index=0,
-        #     key="st_selectbox_llm"
-        # )
+        # LLM model panel
+        llms = list(cfg.llms.available_models)
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply app/frontend/streamlit_app_talk2scholars.py:130]: I have made the changes

File: app/frontend/streamlit_app_talk2scholars.py [hunk 10]
```diff
@@ -214,24 +227,47 @@ def update_llm_model():
...
-                        tracer = LangChainTracer(
-                            project_name=st.session_state.project_name
-                        )
-                        # Get response from the agent
-                        response = app.invoke(
-                            {"messages": [HumanMessage(content=prompt)]},
-                            config=config | {"callbacks": [tracer]},
-                        )
-                        st.session_state.run_id = cb.traced_runs[-1].id
+                    # with collect_runs() as cb:
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply app/frontend/streamlit_app_talk2scholars.py:240]: Thanks for the feedback! I’ve cleaned up unnecessary commented-out code and ensured Langsmith tracing is properly implemented, including assigning st.session_state.run_id within collect_runs(). I also reviewed the remaining comments and kept only those that provide meaningful context.
Let me know if this looks good or if any further adjustments are needed! Regarding the Langsmith tracer, I’ve added it in this PR as suggested, but I’m happy to make any refinements based on feedback from @gurdeep330.

File: aiagents4pharma/talk2scholars/configs/tools/multi_paper_recommendation/default.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmccloskey) [reply aiagents4pharma/talk2scholars/configs/tools/multi_paper_recommendation/default.yaml]: OK. Makes sense. Thanks for adding the additional comment so that I won't ask about it in a future PR 😊

File: aiagents4pharma/talk2scholars/tools/s2/multi_paper_rec.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dmccloskey) [reply aiagents4pharma/talk2scholars/tools/s2/multi_paper_rec.py:122]: Gotcha 👍

File: aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dmccloskey) [reply aiagents4pharma/talk2scholars/tools/s2/single_paper_rec.py:118]: Sounds good.

**Review by dmccloskey — APPROVED** [review]
All clear 😀
